sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextRowCountEstimator.java

```diff
@@ -0,0 +1,154 @@
+  private static final long DEFAULT_NUM_BYTES_LINES = 1024L;
+  private static final Compression DEFAULT_COMPRESSION = Compression.AUTO;
+  private static final FileIO.ReadMatches.DirectoryTreatment DEFAULT_DIRECTORY_TREATMENT =
+      FileIO.ReadMatches.DirectoryTreatment.SKIP;
+  private static final EmptyMatchTreatment DEFAULT_EMPTY_MATCH_TREATMENT =
+      EmptyMatchTreatment.DISALLOW;
+
+  public abstract int getNumSampledFiles();
+
+  public boolean isSamplingAllFiles() {
+    return getNumSampledFiles() == -1;
+  }
+
+  public abstract long getNumSampledBytes();
+
+  @Nullable
+  @SuppressWarnings("mutable")
+  public abstract byte[] getDelimiters();
+
+  public abstract String getFilePattern();
+
+  public abstract Compression getCompression();
+
+  public abstract EmptyMatchTreatment getEmptyMatchTreatment();
+
+  public abstract FileIO.ReadMatches.DirectoryTreatment getDirectoryTreatment();
+
+  public static TextRowCountEstimator.Builder builder() {
+    return (new AutoValue_TextRowCountEstimator.Builder())
+        .sampleAllFiles()
+        .setNumSampledBytes(DEFAULT_NUM_BYTES_LINES)
+        .setCompression(DEFAULT_COMPRESSION)
+        .setDirectoryTreatment(DEFAULT_DIRECTORY_TREATMENT)
+        .setEmptyMatchTreatment(DEFAULT_EMPTY_MATCH_TREATMENT);
+  }
+
+  /** Builder for {@link org.apache.beam.sdk.io.TextRowCountEstimator}. */
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    public abstract Builder setNumSampledFiles(int numSampledFiles);
+
+    public abstract Builder setNumSampledBytes(long numSampledBytes);
+
+    public abstract Builder setDirectoryTreatment(
+        FileIO.ReadMatches.DirectoryTreatment directoryTreatment);
+
+    public abstract Builder setCompression(Compression compression);
+
+    public abstract Builder setDelimiters(byte[] delimiters);
+
+    public abstract Builder setFilePattern(String filePattern);
+
+    public abstract Builder setEmptyMatchTreatment(EmptyMatchTreatment emptyMatchTreatment);
+
+    public Builder sampleAllFiles() {
+      return this.setNumSampledFiles(-1);
+    }
+
+    public abstract TextRowCountEstimator build();
+  }
+
+  /**
+   * Estimates the number of non empty rows. Note that it does not cache the previous estimate. It
+   * returns 0L if all the lines that have been read are empty.
+   *
+   * @return Number of estimated rows.
+   * @throws IOException
+   */
+  public Long estimateRowCount(PipelineOptions pipelineOptions) throws IOException {
+    long linesSize = 0;
+    int numberOfReadLines = 0;
+    long totalFileSizes = 0;
+    int numberOfReadFiles = 0;
+
+    MatchResult match = FileSystems.match(getFilePattern(), getEmptyMatchTreatment());
+
+    for (MatchResult.Metadata metadata : match.metadata()) {
+
+      if ((!isSamplingAllFiles()) && numberOfReadFiles > getNumSampledFiles()) {
+        break;
+      }
+
+      FileIO.ReadableFile file =
+          FileIO.ReadMatches.matchToReadableFile(
```

Comment:
I am not sure if `directoryTreatment` matters here, we probably can just always skip them. It might matter for actual processing but for row count estimation we have to look at the files no matter what kind directory handling we want at run time.

Reply:
Yes. I have set it as a default value to SKIP. So it will just skip the directories. Do you think it is better to remove the setter for that?